zebra-state/src/service/chain_tip/tests/vectors.rs

```diff
@@ -39,6 +39,13 @@ fn chain_tip_change_is_initially_not_ready() {
     let (_chain_tip_sender, _latest_chain_tip, mut chain_tip_change) =
         ChainTipSender::new(None, Mainnet);
 
+    // TODO: use `unconstrained` to avoid spurious waits from tokio's cooperative multitasking
+    //       (needs a recent tokio version)
```

Comment:
Question: I'm not sure I understood the reason. Why does it need to use `unconstrained` here? This test doesn't use the Tokio runtime, right?

Reply:
Here's my best guess at what's happening here:
* [`FuturesExt::now_or_never`](https://docs.rs/futures/0.3.17/futures/future/trait.FutureExt.html#method.now_or_never) runs the `wait_for_tip_change` future without an executor, [using a no-op waker](https://docs.rs/futures-util/0.3.17/src/futures_util/future/future/mod.rs.html#596-609)
* `wait_for_tip_change` calls [`watch::Receiver::changed`](https://docs.rs/tokio/0.3.7/tokio/sync/watch/struct.Receiver.html#method.changed), which calls some [low-level tokio methods](https://docs.rs/tokio/0.3.7/src/tokio/sync/watch.rs.html#243-257
)
* these low-level methods [sometimes yield to the executor, even if they have more work to do](https://docs.rs/tokio/1.11.0/tokio/task/index.html#cooperative-scheduling)
* but there probably aren't enough tokio operators to trigger one of those yield points ([the original limit was 128](https://tokio.rs/blog/2020-04-preemption))
* and the test is checking that there is no initial change in a `watch::Receiver`, so it passes even if tokio yields early

If we use [tokop::task::unconstrained](https://docs.rs/tokio/1.12.0/tokio/task/fn.unconstrained.html), we can ensure we're seeing the underlying `watch::Receiver::changed` wait point, rather than tokio's cooperative yields.

Reply:
I qualified the reference to `unconstrained` in 6fe9907a5.

Reply:
I'm still not sure how that could happen. In my mental model, what you described would lead to the whole `wait_for_tip_change` future returning `Poll::Pending` (i.e., that's how it would yield early), so the test would fail if that happened because of the `expect` :thinking:

Reply:
This code is perhaps not as clear as it could be.

```rust
    let first = chain_tip_change
        .wait_for_tip_change()
        .now_or_never()
        .transpose()
        .expect("watch sender is not dropped");

    assert_eq!(first, None);
```

Since we're using `transpose`, the `expect` is called on `Result<_, watch::error::RecvError>`, which is `Err(_)` if the sender is dropped. (Which should never happen, because the sender is bound by `chain_tip_change` until the end of the test.)

Then `assert_eq!(first, None)` checks that the future did not resolve immediately, which can happen if:
- the sender never changed, or
- tokio's cooperative multitasking returned `Poll::Pending` early.

Reply:
Oh, now I got it! Thanks! :grin: